

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>6. The combine command &mdash; snpToolkit  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="5. The viz command" href="5_The_viz_command.html" />
    <link href="_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> snpToolkit
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_How_to_Install.html">1. How to Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_snpToolkit_menu.html">2. snpToolkit menu</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_The_explore_command.html">3. The explore command</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_The_annotate_command.html">4. The annotate command</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_The_viz_command.html">5. The viz command</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">6. The combine command</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#options">6.1. Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-the-combine-command">6.2. Running the combine command</a></li>
<li class="toctree-l2"><a class="reference internal" href="#find-and-include-missing-data">6.3. Find and Include Missing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#excluding-snps">6.4. Excluding SNPs</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">snpToolkit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">6. </span>The combine command</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/6_The_combine_command.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="the-combine-command">
<h1><span class="section-number">6. </span>The combine command<a class="headerlink" href="#the-combine-command" title="Permalink to this headline">¶</a></h1>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ snptoolkit combine  -h
usage: snptoolkit combine <span class="o">[</span>-h<span class="o">]</span> --loc LOCATION <span class="o">[</span>-r RATIO<span class="o">]</span> <span class="o">[</span>--bam BAMFILTER BAMFILTER BAMFILTER<span class="o">]</span> <span class="o">[</span>--snps <span class="o">{</span>ns,s,all,inter<span class="o">}]</span> <span class="o">[</span>-e EXCLUDE<span class="o">]</span>

optional arguments:
  -h, --help            show this <span class="nb">help</span> message and <span class="nb">exit</span>

snpToolkit combine required options:
  --loc LOCATION   provide <span class="k">for</span> example the name of the chromosome or plasmid you want to create fasta alignemnt <span class="k">for</span>

snpToolkit additional options:
  -r RATIO              new versus reference allele ratio to filter SNPs from snpToolkit outputs. default <span class="o">[</span><span class="m">0</span><span class="o">]</span>
  --bam BAMFILTER BAMFILTER BAMFILTER
                        provide the depth, ratio and the path to the folder containing the bam files. eg. <span class="m">3</span> <span class="m">0</span>.9 path
  --snps <span class="o">{</span>ns,s,all,inter<span class="o">}</span>
                        Specify <span class="k">if</span> you want to concatenate all SNPs or just synonymous <span class="o">(</span>s<span class="o">)</span>, non-synonymous <span class="o">(</span>ns<span class="o">)</span> or intergenic <span class="o">(</span>inter<span class="o">)</span> SNPs. default <span class="o">[</span>all<span class="o">]</span>
  -e EXCLUDE            Provide a yaml file with keywords and coordinates to be excluded
</pre></div>
</div>
<div class="section" id="options">
<h2><span class="section-number">6.1. </span>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 13%" />
<col style="width: 87%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>–loc</strong></p></td>
<td><p>The name of chromosome or plasmid you want to concatenate the SNPs for. This can be found in the last coloumn of the output file of the annotate command</p></td>
</tr>
</tbody>
</table>
<p>Several options are additional:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>–bam</strong></p></td>
<td><p>This option takes three parameters in the following order: depth ration path_to_bam_files. See below for more details</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–snps</strong></p></td>
<td><p>Type of SNPs to be concatenated. default [all]</p></td>
</tr>
<tr class="row-even"><td><p><strong>-r</strong></p></td>
<td><p>r = M/M+R  where <strong>M</strong> is the number of reads that carry the mutated allele and <strong>R</strong> is the is the number of reads that carry the reference allele. If not specified all SNPs will be taken into account.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-e</strong></p></td>
<td><p>This is to specify a yaml file with two arguments KEYWORDS and COORDINATES. See below for more details</p></td>
</tr>
</tbody>
</table>
<p>The command combine should be run in the directory containing the snpToolkit output files generated using the annotate command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>├── sample1_snpToolkit_SNPs.txt
├── sample3_snpToolkit_SNPs.txt
├── sample5_snpToolkit_SNPs.txt
├── sample9_snpToolkit_SNPs.txt
├── sample10_snpToolkit_SNPs.txt
├── sample2_snpToolkit_SNPs.txt
├── sample4_snpToolkit_SNPs.txt
├── sample6_snpToolkit_SNPs.txt
├── sample7_snpToolkit_SNPs.txt
├── sample8_snpToolkit_SNPs.txt
</pre></div>
</div>
</div>
<div class="section" id="running-the-combine-command">
<h2><span class="section-number">6.2. </span>Running the combine command<a class="headerlink" href="#running-the-combine-command" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ snptoolkit combine  --loc NC_003143.1
<span class="o">[</span><span class="m">09</span>:33:38<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Searching <span class="k">for</span> polymorphic sites...<span class="o">]</span>
<span class="o">[</span><span class="m">09</span>:33:38<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>SNPs polymorphic sites distribution. Please wait...<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">##############################################################################| 490/490 [00:00&lt;00:00, 20233.61it/s]</span>
<span class="o">[</span><span class="m">09</span>:33:38<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_polymorphic_sites.txt<span class="o">]</span>
<span class="o">[</span><span class="m">09</span>:33:38<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_alignment.fasta<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">##############################################################################| 10/10 [00:00&lt;00:00, 4712.17it/s]</span>
</pre></div>
</div>
<p>In the command above, the first step is to search for all polymorphic sites. A total number of 490 SNPs was found. As we didnt use the -r option, all SNPs were analyzed. The minimum ratio in this case will be 0.9</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is important to remember that these snpToolkit output files were generated with the following command:</p>
<p>snptoolkit annotate -i vcf -g GCF_000009065.1_ASM906v1_genomic.gbff -d 5 -q 30 -r 0.9 -p 4</p>
<p>all annotated SNPs  will have at least a depth of 5 and a ratio of 0.9.</p>
</div>
<p>Now lets run the command above with the option -r 1</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ snptoolkit combine  --loc NC_003143.1 -r <span class="m">1</span>
<span class="o">[</span><span class="m">10</span>:06:27<span class="o">]</span> <span class="o">[</span>WARNING<span class="o">]</span> <span class="o">[</span>SNPs_polymorphic_sites.txt exists already and was created on Thu Oct <span class="m">22</span> <span class="m">09</span>:44:43 <span class="m">2020</span>. This file will be replaced.
                                            Press any key to <span class="k">continue</span> or ctrl-c to exit!<span class="o">]</span>
<span class="o">[</span><span class="m">10</span>:06:28<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Searching <span class="k">for</span> polymorphic sites...<span class="o">]</span>
<span class="o">[</span><span class="m">10</span>:06:28<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>SNPs polymorphic sites distribution. Please wait...<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">#################################################################################| 470/470 [00:00&lt;00:00, 22782.49it/s]</span>
<span class="o">[</span><span class="m">10</span>:06:28<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_polymorphic_sites.txt<span class="o">]</span>
<span class="o">[</span><span class="m">10</span>:06:28<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_alignment.fasta<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">#################################################################################| 10/10 [00:00&lt;00:00, 9736.08it/s]</span>
</pre></div>
</div>
<p>As the file SNPs_polymorphic_sites.txt exists already, snpToolkit warn you that you need to change the file name or it will be replaced by the new output file.</p>
<p>As we requested that for all SNPs, 100% of the reads must have the new allele, the number of polymorphic sites is now 470.</p>
<p>The Polymorphic sites output SNPs_polymorphic_sites.txt is as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">##snpToolkit=version</span>
<span class="c1">##commandline= snptoolkit combine --loc NC_003143.1 -r 1</span>
<span class="c1">##location=NC_003143.1</span>
<span class="c1">##Number of polymorphic sites= 470</span>
<span class="c1">##ID    Coordinates     REF     SNP     Location        Product Orientation     NucPosition     REF-codon       NEW-codon       REF-AA  NEW-AA  ProPostion      Type    sample10        sample9 sample8 sample7 sample6 sample5 sample4 sample2 sample3 sample1</span>
snp1    <span class="m">82</span>      C       A       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp2    <span class="m">130</span>     G       C       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp3    <span class="m">855</span>     G       A       YPO_RS01010<span class="p">|</span>asnC        transcriptional regulator AsnC  -       <span class="m">411</span>     ACC     AC<span class="o">[</span>T<span class="o">]</span>   T       T       <span class="m">137</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp4    <span class="m">18061</span>   C       T       YPO_RS01090<span class="p">|</span>YPO_RS01090 IS256 family transposase        +       <span class="m">156</span>     AAC     AA<span class="o">[</span>T<span class="o">]</span>   N       N       <span class="m">52</span>      Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp5    <span class="m">21219</span>   C       A       YPO_RS01110<span class="p">|</span>YPO_RS01110 serine/threonine protein kinase +       <span class="m">428</span>     GCC     G<span class="o">[</span>A<span class="o">]</span>C   A       D       <span class="m">143</span>     NS      <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp6    <span class="m">42303</span>   C       T       YPO_RS01190<span class="p">|</span>fabY        fatty acid biosynthesis protein FabY    +       <span class="m">897</span>     GTC     GT<span class="o">[</span>T<span class="o">]</span>   V       V       <span class="m">299</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>
snp7    <span class="m">61685</span>   G       C       intergenic      .       +       <span class="m">64</span> bp from YPO_RS01280<span class="p">|</span>YPO_RS01280      .       .       .       .       .       .       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp8    <span class="m">74539</span>   C       T       YPO_RS01350<span class="p">|</span>envC        murein hydrolase activator EnvC -       <span class="m">361</span>     GCC     <span class="o">[</span>A<span class="o">]</span>CC   A       T       <span class="m">121</span>     NS      <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp9    <span class="m">76590</span>   C       T       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>
snp10   <span class="m">90931</span>   T       A       YPO_RS01440<span class="p">|</span>YPO_RS01440 molybdate ABC transporter substrate-binding protein     -       <span class="m">578</span>     CAG     C<span class="o">[</span>T<span class="o">]</span>G   Q       L       <span class="m">193</span>     NS      <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
</pre></div>
</div>
<p>The first lines of this file contain a summary and useful information. The SNPs annotation is organized in tab delimited table. The columns of this table are:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 76%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ID</p></td>
<td><p>Identifiier of the SNP</p></td>
</tr>
<tr class="row-odd"><td><p>Coordinates</p></td>
<td><p>SNP coordinate</p></td>
</tr>
<tr class="row-even"><td><p>REF</p></td>
<td><p>Reference allele</p></td>
</tr>
<tr class="row-odd"><td><p>SNP</p></td>
<td><p>New allele in analyzed sample</p></td>
</tr>
<tr class="row-even"><td><p>Locatio</p></td>
<td><p>location within the genome</p></td>
</tr>
<tr class="row-odd"><td><p>Product</p></td>
<td><p>Functional product of the gene</p></td>
</tr>
<tr class="row-even"><td><p>Orientation</p></td>
<td><p>Gene orientation</p></td>
</tr>
<tr class="row-odd"><td><p>NucPosition</p></td>
<td><p>gene Coordinate of the SNP within the gene</p></td>
</tr>
<tr class="row-even"><td><p>REF-codon</p></td>
<td><p>Reference codon</p></td>
</tr>
<tr class="row-odd"><td><p>NEW-codon</p></td>
<td><p>New codon</p></td>
</tr>
<tr class="row-even"><td><p>Ref AA</p></td>
<td><p>Amino Acid corresponding to reference codon</p></td>
</tr>
<tr class="row-odd"><td><p>SNP AA</p></td>
<td><p>Amino Acid corresponding to new codon</p></td>
</tr>
<tr class="row-even"><td><p>ProPostion</p></td>
<td><p>Coordinate of the Amino Acid</p></td>
</tr>
<tr class="row-odd"><td><p>Type</p></td>
<td><p>Could be Synonymous (Syn) or Non-Synonymous (NS), or (.) for intergenic</p></td>
</tr>
</tbody>
</table>
<p>After these columns, each column will represented one analyzed sample. The presence or absence of each SNP is represented by 1 or 0, respectively.</p>
<p>In addition to the SNPs_polymorphic_sites.txt, snpToolkit will also generates a fasta file SNPs_alignment.fasta containing the concatenation of all polymorphic sites on each sample.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ grep <span class="s1">&#39;&gt;&#39;</span> SNPs_alignment.fasta
&gt;NC_003143.1
&gt;sample10
&gt;sample9
&gt;sample8
&gt;sample7
&gt;sample6
&gt;sample5
&gt;sample4
&gt;sample2
&gt;sample3
&gt;sample1
</pre></div>
</div>
<p>The first sequence is the reference sequence followed by the 10 samples used for this example</p>
</div>
<div class="section" id="find-and-include-missing-data">
<h2><span class="section-number">6.3. </span>Find and Include Missing data<a class="headerlink" href="#find-and-include-missing-data" title="Permalink to this headline">¶</a></h2>
<p>Lets now suppose that we have two ancient DNA samples that we have analyzed and generated the corresponding vcf files. When working with aDNA, usually not 100% of your genome is recovered. When looking for the distribution of all polymorphic sites within these aDNA, it is important to know if an SNP was not identified because for that position the aDNA is similar to the reference or because the region is not covered at all. To be able to identify such position, users have to provide the bam files of all samples for whom  they want to account for missing data.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snptoolkit combine -r <span class="m">0</span>.9 --loc NC_003143.1 --bam <span class="m">2</span> <span class="m">1</span>.0 ../bam/
</pre></div>
</div>
<p>As you can see, you need just to specify one addition option ‘- -bam’ with three parameter</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>--bam <span class="m">2</span> <span class="m">1</span>.0 ../bam/
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 13%" />
<col style="width: 87%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>Minimum depth of coverage to consider. Here this depth is set to 2, which mean that at least 2 reads should be found</p></td>
</tr>
<tr class="row-odd"><td><p>1.0</p></td>
<td><p>Number of reads with new allele /  total number of reads ratio. For this example it is set to 1.0, which mean that the 2 reads should have the new allele</p></td>
</tr>
<tr class="row-even"><td><p>../bam</p></td>
<td><p>This is the path of the folder containing the bam files of all aDNA to be considered. The bam folder should also include the .bai files</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When using the –bam option, snptoolkit will create in total 4 output files:</p>
<ul class="simple">
<li><p>SNPs_polymorphic_sites.txt + SNPs_alignment.fasta</p></li>
<li><p>SNPs_polymorphic_sites_clean.txt + SNPs_alignment.fasta</p></li>
</ul>
<p>As described above, The first two files contains all SNPs found in all analysed samples including polymorphic sites where in some samples there is missing information indicated by a question mark.
The second two files are a “clean” version of the two files described above in the sence that they don’t contain any position where missing information is reported.</p>
</div>
<a class="reference external image-reference" href="6_The_combine_command/Figure7.png"><img alt="6_The_combine_command/Figure7.png" src="_images/Figure7.png" /></a>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>├── <span class="o">[</span> 19M <span class="o">]</span>  sampleY.bam
├── <span class="o">[</span> 33M <span class="o">]</span>  sampleX.bam
├── <span class="o">[</span> 14K <span class="o">]</span>  sampleX.bam.bai
└── <span class="o">[</span><span class="m">6</span>.6K <span class="o">]</span>  sampleY.bam.bai
</pre></div>
</div>
<p>Now lets run the combine command on all snpToolkit output files generated using the annotate command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>├── <span class="o">[</span> 10K <span class="o">]</span>  sample9_snpToolkit_SNPs.txt
├── <span class="o">[</span> 10K <span class="o">]</span>  sampleY_snpToolkit_SNPs.txt
├── <span class="o">[</span> 32K <span class="o">]</span>  sample10_snpToolkit_SNPs.txt
├── <span class="o">[</span> 15K <span class="o">]</span>  sample1_snpToolkit_SNPs.txt
├── <span class="o">[</span> 15K <span class="o">]</span>  sample2_snpToolkit_SNPs.txt
├── <span class="o">[</span> 12K <span class="o">]</span>  sample3_snpToolkit_SNPs.txt
├── <span class="o">[</span> 35K <span class="o">]</span>  sample4_snpToolkit_SNPs.txt
├── <span class="o">[</span> 36K <span class="o">]</span>  sample5_snpToolkit_SNPs.txt
├── <span class="o">[</span> 38K <span class="o">]</span>  sample6_snpToolkit_SNPs.txt
├── <span class="o">[</span> 37K <span class="o">]</span>  sample7_snpToolkit_SNPs.txt
├── <span class="o">[</span> 16K <span class="o">]</span>  sampleX_snpToolkit_SNPs.txt
├── <span class="o">[</span> 41K <span class="o">]</span>  sample8_snpToolkit_SNPs.txt
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ snptoolkit combine -r <span class="m">0</span>.9 --loc NC_003143.1 --bam <span class="m">2</span> <span class="m">1</span>.0 ../bam/
<span class="o">[</span><span class="m">10</span>:45:48<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Searching <span class="k">for</span> polymorphic sites...<span class="o">]</span>
<span class="o">[</span><span class="m">10</span>:45:48<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>SNPs polymorphic sites distribution. Please wait...<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">############################################################################################################################################################################################| 505/505 [00:04&lt;00:00, 112.91it/s]</span>
<span class="o">[</span><span class="m">10</span>:45:52<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_alignment.fasta<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">#############################################################################################################################################################################################| 12/12 [00:00&lt;00:00, 8558.35it/s]</span>
<span class="o">[</span><span class="m">10</span>:45:52<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_polymorphic_sites_clean.txt<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">#########################################################################################################################################################################################| 375/375 [00:00&lt;00:00, 183381.60it/s]</span>
<span class="o">[</span><span class="m">10</span>:45:52<span class="o">]</span> <span class="o">[</span>INFO<span class="o">]</span> <span class="o">[</span>Creating SNPs_alignment_clean.fasta<span class="o">]</span>
progress: <span class="m">100</span>%<span class="p">|</span><span class="c1">############################################################################################################################################################################################| 12/12 [00:00&lt;00:00, 12738.96it/s]</span>
</pre></div>
</div>
<p>By adding the two aDNA samples, the number of polymorphic sites has increased to 505.  The new SNPs_polymorphic_sites.txt contains now the SNPs distribution for sampleX and sampleY.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">##commandline= snptoolkit combine -r 0.9 --loc NC_003143.1 --bam 2 1.0 ../bam/</span>
<span class="c1">##location=NC_003143.1</span>
<span class="c1">##Number of polymorphic sites= 505</span>
<span class="c1">##ID    Coordinates     REF     SNP     Location        Product Orientation     NucPosition     REF-codon       NEW-codon       REF-AA  NEW-AA  ProPostion      Type    sample10        sampleX sampleY sample9 sample8 sample7 sample6 sample5 sample4 sample2 sample3 sample1</span>
snp1    <span class="m">82</span>      C       A       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp2    <span class="m">130</span>     G       C       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp3    <span class="m">855</span>     G       A       YPO_RS01010<span class="p">|</span>asnC        transcriptional regulator AsnC  -       <span class="m">411</span>     ACC     AC<span class="o">[</span>T<span class="o">]</span>   T       T       <span class="m">137</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp4    <span class="m">18061</span>   C       T       YPO_RS01090<span class="p">|</span>YPO_RS01090 IS256 family transposase        +       <span class="m">156</span>     AAC     AA<span class="o">[</span>T<span class="o">]</span>   N       N       <span class="m">52</span>      Syn     <span class="m">0</span>       ?       ?       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp5    <span class="m">21219</span>   C       A       YPO_RS01110<span class="p">|</span>YPO_RS01110 serine/threonine protein kinase +       <span class="m">428</span>     GCC     G<span class="o">[</span>A<span class="o">]</span>C   A       D       <span class="m">143</span>     NS      <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp6    <span class="m">29368</span>   G       T       YPO_RS01140<span class="p">|</span>hemN        oxygen-independent coproporphyrinogen III oxidase       +       <span class="m">387</span>     GTG     GT<span class="o">[</span>T<span class="o">]</span>   V       V       <span class="m">129</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp7    <span class="m">42303</span>   C       T       YPO_RS01190<span class="p">|</span>fabY        fatty acid biosynthesis protein FabY    +       <span class="m">897</span>     GTC     GT<span class="o">[</span>T<span class="o">]</span>   V       V       <span class="m">299</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>
snp8    <span class="m">61685</span>   G       C       intergenic      .       +       <span class="m">64</span> bp from YPO_RS01280<span class="p">|</span>YPO_RS01280      .       .       .       .       .       .       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp9    <span class="m">74539</span>   C       T       YPO_RS01350<span class="p">|</span>envC        murein hydrolase activator EnvC -       <span class="m">361</span>     GCC     <span class="o">[</span>A<span class="o">]</span>CC   A       T       <span class="m">121</span>     NS      <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
</pre></div>
</div>
<p>For snp4, this SNP is considered as “?” as at position 18061 the criteria minimum 2 reads AND ratio 1.0 were not satisfied</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">##ID    Coordinates     REF     SNP     Location        Product Orientation     NucPosition     REF-codon       NEW-codon       REF-AA  NEW-AA  ProPostion      Type    sample10        sampleX sampleY sample9 sample8 sample7 sample6 sample5 sample4 sample2 sample3 sample1</span>
snp4    <span class="m">18061</span>   C       T       YPO_RS01090<span class="p">|</span>YPO_RS01090 IS256 family transposase        +       <span class="m">156</span>     AAC     AA<span class="o">[</span>T<span class="o">]</span>   N       N       <span class="m">52</span>      Syn     <span class="m">0</span>       ?       ?       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
</pre></div>
</div>
<p>Lets take a look now at the file SNPs_alignment.fasta:</p>
<a class="reference external image-reference" href="6_The_combine_command/Figure6.png"><img alt="6_The_combine_command/Figure6.png" src="_images/Figure6.png" /></a>
<p>The file SNPs_polymorphic_sites_clean.txt contains only 375 SNPs instead of 505 as 130 polymorphic sites contain missing information.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">##commandline= snptoolkit combine -r 0.9 --loc NC_003143.1 --bam 2 1.0 ../bam/</span>
<span class="c1">##location=NC_003143.1</span>
<span class="c1">##Number of polymorphic sites= 375</span>
<span class="c1">##ID    Coordinates     REF     SNP     Location        Product Orientation     NucPosition     REF-codon       NEW-codon       REF-AA  NEW-AA  ProPostion      Type    sample10        sampleX sampleY sample9 sample8 sample7 sample6 sample5 sample4 sample2 sample3 sample1</span>
snp1    <span class="m">82</span>      C       A       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp2    <span class="m">130</span>     G       C       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>
snp3    <span class="m">855</span>     G       A       YPO_RS01010<span class="p">|</span>asnC        transcriptional regulator AsnC  -       <span class="m">411</span>     ACC     AC<span class="o">[</span>T<span class="o">]</span>   T       T       <span class="m">137</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span> <span class="m">0</span>
snp4    <span class="m">21219</span>   C       A       YPO_RS01110<span class="p">|</span>YPO_RS01110 serine/threonine protein kinase +       <span class="m">428</span>     GCC     G<span class="o">[</span>A<span class="o">]</span>C   A       D       <span class="m">143</span>     NS      <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span> <span class="m">0</span>
snp5    <span class="m">29368</span>   G       T       YPO_RS01140<span class="p">|</span>hemN        oxygen-independent coproporphyrinogen III oxidase       +       <span class="m">387</span>     GTG     GT<span class="o">[</span>T<span class="o">]</span>   V       V       <span class="m">129</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span> <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>
snp6    <span class="m">42303</span>   C       T       YPO_RS01190<span class="p">|</span>fabY        fatty acid biosynthesis protein FabY    +       <span class="m">897</span>     GTC     GT<span class="o">[</span>T<span class="o">]</span>   V       V       <span class="m">299</span>     Syn     <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span> <span class="m">1</span>       <span class="m">0</span>
snp7    <span class="m">61685</span>   G       C       intergenic      .       +       <span class="m">64</span> bp from YPO_RS01280<span class="p">|</span>YPO_RS01280      .       .       .       .       .       .       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>       <span class="m">0</span> <span class="m">0</span>
snp8    <span class="m">74539</span>   C       T       YPO_RS01350<span class="p">|</span>envC        murein hydrolase activator EnvC -       <span class="m">361</span>     GCC     <span class="o">[</span>A<span class="o">]</span>CC   A       T       <span class="m">121</span>     NS      <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span>       <span class="m">1</span> <span class="m">1</span>
snp9    <span class="m">76590</span>   C       T       intergenic      .       +       .       -       -       -       -       -       -       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">0</span>       <span class="m">1</span>       <span class="m">0</span>
</pre></div>
</div>
<p>Lets take a look now at the file SNPs_alignment_clean.fasta that compared to SNPs_alignment.fasta does not contain polymorphic site with missing information.</p>
<a class="reference external image-reference" href="6_The_combine_command/Figure8.png"><img alt="6_The_combine_command/Figure8.png" src="_images/Figure8.png" /></a>
</div>
<div class="section" id="excluding-snps">
<h2><span class="section-number">6.4. </span>Excluding SNPs<a class="headerlink" href="#excluding-snps" title="Permalink to this headline">¶</a></h2>
<p>For the combine command it is possible also to provide a file that includes the SNPs to exclude with the option -e when searching for all polymorphic sites and building the fasta file. Compared to the annotate command, the file here must be in yaml format and not a tabulated format. Here is an example:</p>
<blockquote>
<div><p>KEYWORDS: YPO3746;YPO3747</p>
<p>COORDINATES: 64265;7662</p>
</div></blockquote>
<p>The yaml file has two arguments: KEYWORDS and COORDINATES. With the argument KEYWORDS you can specify genes names, a word within a product name, etc…</p>
<ul class="simple">
<li><p>If you specify a gene name, all SNPs within that gene will be excluded.</p></li>
<li><p>If you specify for example ‘cytochrome’, all genes having the word cytochrome in their product will be ignored.</p></li>
</ul>
<p>For the argument COORDINATES, it is straightforward, it will contains the list of SNP positions to be excluded.</p>
<p>The rational behind this choice of file structure is that after creating the files SNPs_polymorphic_sites.txt and SNPs_alignment.fasta, and making some downstream analyses like building phylogenetic trees and checking some SNPs for confirmation, it is possible to exclude for example SNPs that are in regions that included some repeats or a family of genes that has duplicated genes… such file will give all the flexibility for users.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the next version of snpToolkit, it will be possible instead of providing the bam files, to search back the original vcf files if only they include all the positions and not only variable sites:</p>
<ul class="simple">
<li><p>With samtools mpileup you can use the option -aa to output all positions, including unused reference sequences.</p></li>
<li><p>With gatk haplotypeCaller you can use mode EMIT_ALL_SITES  with the option -output-mode</p></li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="5_The_viz_command.html" class="btn btn-neutral float-left" title="5. The viz command" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Amine Namouchi

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>